### ISSUE TITLE
Allow using all list/watch

### ISSUE THREAD

**Issue by brancz** [prometheus-operator/prometheus-operator#3840]
**What is missing?**

Currently we intentionally scope service monitors to the namespace they are in and the generated scrape configs set the namespaces to watch. This results in 1 list/watch to be created per namespace being monitored in Prometheus.

**Why do we need it?**

Prometheus is able to re-use discovered targets if only relabeling configs are different. This could potentially safe a lot of resources spent on maintaining a lot of list/watches.

It's not entirely clear from the data yet, but it's possible that having a lot of ServiceMonitors is putting a lot of strain on Prometheus (RE: https://github.com/prometheus/prometheus/issues/8014). In any case this seems like a trade-off people should be able to choose if they wanted to.

The flip side of this functionality is that Prometheus would need cluster wide list/watch permission (a lot of people already run Prometheus this way so it would just be a net win in terms of less resources used).

**Comment by shaikatz** [comment]
I would like to try and work on that feature.
Can I get some pointers?
If we remove the namespace definition from the generated sd config, what will happen if the user has two targets matching the same labels on two different namespaces? 
Maybe we can just move the namespace to drop relabeling instead?

**Comment by simonpasquier** [comment]
@shaikatz yes the idea is to remove the `namespaces: names: [foo]` part from the generated kubernetes SD config and drop targets that don't match the service monitor's namespace at the relabeling stage.
I'm not sure how we're going to expose this to the users: probably a new field in the Prometheus CRD?

**Comment by shaikatz** [comment]
I think a minimal change without changing user experience is moving the namespace filtering to relabeling stage only if no namespaceSelector was specified.
WDYT?

**Comment by shaikatz** [comment]
I think this setting should be in the operator level, something that will indicate that Prometheus has full cluster list/watch access, and it will make the generated SD config use relabeling instead of SD namespace filtering.

**Comment by brancz** [comment]
> I'm not sure how we're going to expose this to the users: probably a new field in the Prometheus CRD?

Yeah, this makes sense to me. Essentially `sharedListWatch: true`.
